ParticleHandler: Store vector of particles per cell in std::list

kronbichler opened the pull request:
This PR switches the implementation of the particles inside `ParticleHandler` to `std::list<vector of particles on cell>`, with the goal to increase performance as we iterate over particles. This PR aims to be good at the following three issues:
- Iteration through all particles of a collection is now very cheap, on the order of 6-10 instructions, also when switching from one cell to the next. Part of the advantage comes from the fact that we now have a very short function so the compiler will always inline it when optimizing.
- The access direction from a `cell` to the particles on that cell is solved by a `std::vector<std::list<>::iterator>  cells_to_particle_cache`. This gives O(1) access cost from a cell into a list of particles (using 3-5 instructions)
- In order to have fast access from a particle iterator to a cell, we keep a cell iterator alongside the particle vectors, i.e., the actual container type is now `std::list<std::pair<std::vector<Handle>, cell_iterator>`. This is again a simple lookup.

Comments on changes:
- ~~Even though we tried hard to unify all particles into a single field back in summer, I went along to duplicate the storage lists into `owned_particles` and `ghost_particles` again. The unification still exists via the access through `cells_to_particle_cache`. There are a few spots where we look up two containers in (almost) the same way, but I nonetheless prefer this as the increment in the `ParticleAccessor`, because it reduces then clean without any checks.~~ Edit: We still have a unified data structure, it is better this way.
- ~~I still have 16 test failures when I run `ctest -R particle`. Most of them are because we do not have the exact same order through the cells (albeit I sort them in the end) and for particles within a cell.~~ But there is one more spurious error where I believe that either the compiler has a bug (as it only happens with "-O3") or my code has a bug that corrupts the stack (more likely).

FYI @blaisb @gassmoeller @peterrum

gassmoeller commented on include/deal.II/particles/particle_handler.h:
```diff
@@ -902,10 +913,21 @@ namespace Particles
+    particle_container owned_particles;
+
+    /**
+     * Set of particles currently living in the locally owned cells.
```
```suggestion
     * Set of particles currently living in the ghost cells.
```

gassmoeller commented on include/deal.II/particles/particle_handler.h:
```diff
@@ -79,8 +79,9 @@ namespace Particles
      */
-    using particle_container =
-      std::vector<std::vector<typename PropertyPool<dim, spacedim>::Handle>>;
+    using particle_container = std::list<
+      std::pair<std::vector<typename PropertyPool<dim, spacedim>::Handle>,
+                typename Triangulation<dim, spacedim>::active_cell_iterator>>;
```
I wonder if we can define a `struct ParticlesInCell` that contains a `std::vector<typename PropertyPool<dim, spacedim>::Handle> particles;` and a `typename Triangulation<dim, spacedim>::active_cell_iterator> cell_iterator`.
 Then a lot of the calls below would be easier to read. Instead of `particles_in_cell->first/second` it would be `particles_in_cell->particles/cell_iterator`. Do you think that would have a performance impact?

kronbichler replied:
That's a good idea and helps documenting all first/second instances. It won't have any performance impact. I will change in the next few days.

kronbichler replied:
@gassmoeller this is now in f7c21cd

gassmoeller reviewed (commented):
I think the new structure makes sense, we mostly wanted to unify the containers because the old `std::vector` would have a lot of unused overhead for the ghost particles anyway. Separating particles from ghost particles is fine for the `list` containers. I dont have time for a proper review right now, but will take a look until tomorrow.

gassmoeller commented on source/particles/particle_handler.cc:
```diff
@@ -455,26 +475,50 @@ namespace Particles
 
 
+  template <int dim, int spacedim>
+  void
```
would it make sense to let this function return a particle_iterator? The one other function called `insert_particle` does that at the moment.

gassmoeller commented on source/particles/particle_handler.cc:
```diff
@@ -495,20 +539,20 @@ namespace Particles
+    insert_particle(property_pool->register_particle(), cell, owned_particles);
+    const typename particle_container::iterator &cache =
+      cells_to_particle_cache[cell->active_cell_index()];
+    particle_iterator particle_it(owned_particles,
+                                  cache,
                                   *property_pool,
-                                  cell,
-                                  particles[active_cell_index].size() - 1);
+                                  cache->particles.size() - 1);
```
if the function returned an iterator we could save these lines here

gassmoeller commented on source/particles/particle_handler.cc:
```diff
@@ -1709,14 +1761,17 @@ namespace Particles
-          *property_pool,
-          cell,
-          received_particles[active_cell_index].size() - 1);
+        insert_particle(property_pool->register_particle(),
+                        cell,
+                        received_particles);
+        const typename particle_container::iterator &cache =
+          cells_to_particle_cache[cell->active_cell_index()];
+        Assert(cache->cell_iterator == cell, ExcInternalError());
+
+        particle_iterator particle_it(received_particles,
+                                      cache,
+                                      *property_pool,
+                                      cache->particles.size() - 1);
```
This can also be shortened with a returned iterator.

gassmoeller commented on source/particles/particle_handler.cc:
```diff
@@ -2193,20 +2252,25 @@ namespace Particles
 
             // Cannot use range-based loop, because number of particles in cell
             // is going to change
+            if (cache == typename particle_container::iterator())
+              break;
```
This line confuses me, because it should never happen (and if it does something went wrong). Above in line 2216 we inserted the particle data into the new cell. In this case into child(0) of the old cell that was refined. during that operation the cache should have been updated to the correct number of new particles, so cache should never be empty. Maybe make this an Assert?

kronbichler replied:
I don't know exactly why, but when I ran one of the Lethe tests I have the case that `data_range` is empty, and hence, no entries get added. I'm not sure whether this is supposed to happen or not. Either way, I return early from the function in that case now, and changed to an assertion.

gassmoeller commented on source/particles/particle_handler.cc:
```diff
@@ -2222,13 +2286,13 @@ namespace Particles
-                                  particles[stored_index].back();
-                                particles[stored_index].resize(
-                                  particles[stored_index].size() - 1);
+                                cache->particles[i] = cache->particles.back();
+                                cache->particles.resize(
+                                  cache->particles.size() - 1);
```
Assuming we move all particles away from child(0) this will end with an empty but valid vector in the cache for child(0). Is this equivalent to what happens to other cells that have no particles?
Also we may need to erase the list entry in `owned_particles` if that happens. Maybe we should replace these lines with a call to `remove_particle()`?

kronbichler replied:
I thought about it, but it gets more complicated, because we would need to take care of the handle as well (setting it to invalid before we enter `remove_particle()`). And then we need to run the `i` loop in reverse order, as we should not remove the particle as we still keep the index `i` into the array. So for now, I added the step to delete the cell from the list instead.

gassmoeller reviewed (commented):
Very nice! I left some minor comments and some questions about complexity and a potential bug upon refinement (not sure). I think the new structure is better, in particular for your application with many empty cells and very little work on each of the particles. I havent seen anything that should slow down other applications (apart from the comment about sorting particles to be removed).

kronbichler commented on include/deal.II/particles/particle_accessor.h:
```diff
@@ -45,11 +45,44 @@ namespace Particles
+      /**
+       * A vector of particles on a cell.
+       */
+      std::vector<typename PropertyPool<dim, spacedim>::Handle> particles;
```
No, this is the actual storage place of the particle. It would morally belong to `ParticleHandler` where the `particle_container` is actually stored, but moved it into `ParticleAccessor` in order not to touch the order of include files or do some strange forward declaration (`particle_accessor.h` is included by `particle_handler.h`). If you find this confusing, we can move `ParticlesInCell` into a new header file, and make sure to let both files include it. What do you think?

kronbichler commented on include/deal.II/particles/particle_accessor.h:
```diff
@@ -875,7 +885,8 @@ namespace Particles
     const ParticleAccessor<dim, spacedim> &other) const
   {
-    return (property_pool == other.property_pool) && (cell == other.cell) &&
+    return (property_pool == other.property_pool) &&
+           (particles_in_cell == other.particles_in_cell) &&
            (particle_index_within_cell == other.particle_index_within_cell);
```
I reworked the particle iterator so that we only need the `particles_in_cell` (i.e., `std::list<ParticleInCell>::iterator`), while completely skipping the pointers to particles. Regarding the comparisons: I think we need the comparison of iterators, at least for now, because we need to be able to compare with the `past-by-one` iterator, where we cannot access the handle. So I think this is the right way to go, and indeed if anything we are more restrictive if we consider the iterators to match as well.

kronbichler commented on source/particles/particle_handler.cc:
```diff
@@ -388,52 +436,24 @@ namespace Particles
-        particles_on_cell->resize(move_to);
-      }
+    // sort particles in reverse order on each cell to ensure the removal with
+    // the other function is valid
```
I agree that we can catch the case where we do not need to sort. I addressed that in the new version of the code.

kronbichler commented:
Thanks for the comments @gassmoeller - I now addressed most of them, except the ones above where I replied to. I also adjusted the test cases I see fail locally, in order to see if the changes are fine for the CI testers. Note that changes are only due to different orders of particles within cells, not changes in the assignment of particles to cells.

Note that I also switched to store all particles in a single container again, because `_GLIBCXX_DEBUG` does not like if I compare iterators of different `std::list` objects, and this is what I necessarily have to do when I flag cells in one way or the other. This necessitated adding some further elements in the `std::list`, that I call *anchor* entries, in order to distinguish what is a owned particle and what is a ghost. Note that these *anchors* are also nice to additionally remove some checks from the `ParticleAccessor` class, as we now can create `begin()` and `end()` iterators from `ParticleHandler` that point to those anchors when trying to see if the loops are completed. While it is a slight complication to keep track of these entries and not change the order, I think it is worth it outside the `update_cached_numbers()` function :wink:

I also note that the old major changelog we wrote in summer still applies to this patch, so I guess we won't need updates there.

gassmoeller reviewed (commented):
Very nice, I like that we are back to one container. About the anchor entries: That is a great use of the flexibility we have now with the list, but I do not see why we need 3 of them at the moment. Shouldnt one anchor be enough (the one in the middle separating owned from ghost entries)? particles.begin() and particles.end() could be used as normal instead of the other two anchors. Or is there some use where this would not work?

Apart from that I am happy with the changes.

I ran the ASPECT tests with this PR and it seems fine in most cases, but in our test for `copy_from` when iterating over the copied particles I get the following:
```
*** stack smashing detected ***: terminated
[gator-station-1:590563] *** Process received signal ***
[gator-station-1:590563] Signal: Aborted (6)
[gator-station-1:590563] Signal code:  (-6)
[gator-station-1:590563] *** Process received signal ***
[gator-station-1:590563] Signal: Segmentation fault (11)
[gator-station-1:590563] Signal code: Address not mapped (1)
[gator-station-1:590563] Failing at address: 0x71
Segmentation fault (core dumped)
```

Does particle_handler_18 pass in deal.II (thats the deal.II test for that function)?

blaisb commented:
@kronbichler  I tested the PR on our code base and it seems in some calls of sort_particles_into_subdomain_and_cell lead to segmentation fault.
The best illustration of this is this test case (which is not as simple as it should be ) : https://github.com/lethe-cfd/lethe/blob/master/tests/dem/two_particles_multiple_contacts_parallel.cc

I have not been able to identify what is the root cause of this, but I will look into it this weekend. I don't think it's on our side though, but I could be wrong.

kronbichler commented:
@gassmoeller I tried to reproduce with the `particles_handler_copy` of ASPECT and got it failing, but not in the particle handler itself but in `aspect::Simulator<2>::build_advection_preconditioner(aspect::Simulator<2>::AdvectionField const&, dealii::TrilinosWrappers::PreconditionILU&, double) (source/simulator/assembly.cc:847` (somewhere there is a write-past-end-of-array in `Teuchos::ParameterEntry` inside the ILU preconditioner). If I didn't miss anything, the error is also not related to this PR, as I get it with master as well. Since I believe there could be something hidden, I'd be happy if you could let valgrind run over this error (with a debug build) and see what it returns, as I can't neither with gcc-7 nor clang-13.

@blaisb I think I fixed the error on your code and (force-) pushed a fix.

kronbichler commented:
> Shouldnt one anchor be enough (the one in the middle separating owned from ghost entries)? particles.begin() and particles.end() could be used as normal instead of the other two anchors. Or is there some use where this would not work?

I tried this first, but then it gets messy to use `ParticleAccessor::state()` as we cannot dereference the `std::list::iterator` for entries before/past the end of valid entries. In debug mode, this would then seg-fault before we trigger an actual assertion, which I did not like. It might be that I just tried with a solution that is too complicated, but I guess we can always see if we find simplifications in the future.

peterrum commented on source/particles/particle_handler.cc:
```diff
@@ -205,66 +224,158 @@ namespace Particles
-    number_of_locally_owned_particles                  = 0;
-    types::particle_index local_max_particle_index     = 0;
-    types::particle_index local_max_particles_per_cell = 0;
+    TriaActiveIterator<CellAccessor<dim, spacedim>> past_the_end_iterator(
+      triangulation, -1, -1);
```
```suggestion
    const auto past_the_end_iterator = tria->end();
```
?

peterrum reviewed (commented):
Looks good to  me. I left some minor comments.

Let's merge once all tests in Lethe and aspects are running again!

kronbichler commented on source/particles/particle_handler.cc:
```diff
@@ -205,66 +224,158 @@ namespace Particles
+    // collective results
+    number_of_locally_owned_particles = 0;
 
+    types::particle_index result[2] = {};
```
I would if there were an instantiation of the overload of `Utilities::MPI::max()` with `std::array` argument. But right now I would manually have to set the arguments in terms of an `ArrayView` and thus much worse code. So I would propose to leave it out of this PR until we have fixed it elsewhere.

kronbichler commented on source/particles/particle_handler.cc:
```diff
@@ -162,6 +166,15 @@ namespace Particles
+         ++it)
+      ++owned_particles_end;
+
+    for (auto it = particles.begin(); it != particles.end(); ++it)
```
I might be missing some features in C++, but my problem is that I do want to use the actual iterator two lines down (as this is what I want to get out of the loop), not the reference I get from the range-based for loop. Or is there a solution with range-based for loops?

peterrum replied:
You are storing the iterator instead of pointer. That is odd. What is the reason for that?

kronbichler replied:
If I kept a pointer to an entry in the list only, I would have no way to go to the next or previous entry - the underlying reason is that `std::list` is a doubly-linked list without a random access into it, so I really need to additional context provided by the `iterator` class. Or do I misunderstand what you mean by `pointer`? Note that at least on the GCC STL implementation with release context an `iterator` is almost a pointer with `sizeof(std::list::iterator)==8`, but it includes the bi-directional pointers to the previous/next list element. In `DEBUG` mode it is something else, with a lot of context to find out about out-of-range access (`sizeof(std::list::iterator)==40`).

peterrum replied:
I see. Thanks for the explanation!

kronbichler commented on source/particles/particle_handler.cc:
```diff
@@ -205,66 +224,158 @@ namespace Particles
+
+    // check that we only have locally owned particles in the first region of
+    // cells; note that we skip the very first anchor element
+    for (auto it = ++particles.begin(); it != owned_particles_end; ++it)
```
I see two possibilities: Either we set the variable to `const` and only set it in a single place by a `const_cast`, or we use some internal functions `owned_particles_begin`, `owned_particles_end`, `ghost_particles_begin`, `ghost_particles_end` that we use everywhere, except for the initialization. This would limit the access confusion (with the three anchors) to a single place where we retrieved them. Opinions?

kronbichler replied:
I opted for additional `begin()/end()` functions to be used internally. This should streamline use.

blaisb commented:
@kronbichler tested the latest version and everything passes in Lethe. Good job :)!

gassmoeller commented on source/particles/particle_handler.cc:
```diff
@@ -205,66 +227,158 @@ namespace Particles
+    for (unsigned int i = 0; i < 3; ++i)
+      given_particles.emplace_back(
+        std::vector<typename PropertyPool<dim, spacedim>::Handle>(),
+        triangulation->end());
```
If triangulation is a `nullptr` this line dereferences a null pointer (which is undefined behavior), which sometimes succeeds (as apparently in the deal.II tests) and sometimes is detected as a corruption of the stack (as on some of my machines for the ASPECT tests). We should only execute these steps if a triangulation is set, because if it is not set the particle handler is not in a usable state anyways, and `initialize` or `copy_from` have to be called before it is usable. Maybe also not call this function at all in the default constructor, because it will be a no-op?
Also I have not seen a reason to give triangulation as an argument to this function, it will always be the stored triangulation pointer (and that is at least a smart pointer), do you think it makes sense to remove the argument? The comparison would then probably need to be `if (triangulation != SmartPointer<const Triangulation<dim, spacedim>, ParticleHandler<dim, spacedim>>())` because we cannot dereference an uninitialized smart pointer (I havent tried this).

This was one of the more interesting debugging sessions :-)

kronbichler replied:
Thanks a lot for the help, this has indeed been wrong. I removed the argument `triangulation` and construct the invalid iterator there. I still kept support for `triangulation == nullptr`, as I want `ParticleHandler::begin()` to return a valid (empty) range rather than throw an exception with internal error. If we want, we could of course change this again and safe three lines of code.

gassmoeller replied:
That is a good point, and yes the current version is fine. That was the last of my comments, ready from my side.

peterrum reviewed (approved):
Thanks for the changes. @blaisb and @gassmoeller seem to be happy as well. @kronbichler Feel free to merge if you think the code is ready!

kronbichler commented:
I think we are ready - the input of three people and two working applications gives confidence. Thanks to the three of you for the help.
